src/NHibernate/Multi/IQueryBatchItem.cs

```diff
@@ -0,0 +1,66 @@
+using System;
+using System.Collections.Generic;
+using System.Data.Common;
+using NHibernate.Engine;
+using NHibernate.SqlCommand;
+
+namespace NHibernate.Multi
+{
+	/// <summary>
+	/// Interface for wrapping query to be batched by <see cref="IQueryBatch"/>
+	/// </summary>
+	public interface IQueryBatchItem<TResult> : IQueryBatchItem
+	{
+		/// <summary>
+		/// Returns loaded typed results by query.
+		/// Must be called only after <see cref="IQueryBatch.Execute"/>.
+		/// </summary>
+		IList<TResult> GetResults();
+		
+		/// <summary>
+		/// Event is executed after results are loaded by batch.
+		/// Loaded results are provided in action parameter.
+		/// </summary>
+		event Action<IList<TResult>> OnAfterLoad;
```

Comment:
This look more like a callback than like an `event`. An `event` should respect general `event` [guidelines](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/events/how-to-publish-events-that-conform-to-net-framework-guidelines), as does `BeforeBindMapping` and `AfterBindMapping` in NHibernate.

Either made it a simple callback (then maybe renamed `AfterLoadCallback`), or a full blown event.

If kept as an event, it should be renamed `AfterLoad` by the way. The `On` prefix is normally reserved for the class method responsible of raising the event if it has been affected, like `OnBeforeBindMapping` and `OnAfterBindMapping` in NHibernate.

I wonder if the naming `AfterLoad` should be changed for avoiding any confusion with the `ISession.Load`, with `ILoadEventListener`/`IPostLoadEventListener`/`IPreLoadEventListener` (which are not about a .Net `event` in spite of their name; this is some Hibernate Java legacy), and with the `ILifeCycle.OnLoad`. Maybe `AfterExecution` or `Executed`, but well, I have not set up my mind on this subject.

About the choice between callback or event, it sounds more natural to me to change it to a simple callback. Then why not re-adding additional `Add` `IQueryBatch` extension overloads for each query interface, like `IQueryBatch Add<TResult>(this IQueryBatch batch, IQueryOver query, Action<IList<TResult>> afterLoad)`. (Doing that with an `event` would look quite unnatural to me.)

Reply:
I'm not sure I understand the problem about event vs callback. I meant it as simple callback. But it was marked as event in interface just for protection. So caller code can't trigger that callback.

Reply:
Then replace it by a setter function if this is not meant to be an `event` and wants to protect against someone calling it from outside. (I do not really see the point of that protection by the way. If the user wants to do moot things, good for him. Anyway he is the one having supplied the callback in the first place, he can still hold on it and call it directly if he wishes.)

Reply:
>Then replace it by a setter function

I don't see a problem being it event or delegate. Let it be whatever you think is best - so I've already made changes you requested.

Reply:
> I don't see a problem being it event or delegate.

Me neither, provided it is an actual event respecting .Net guidelines.

> But it was marked as event in interface just for protection.

In my view, this is a debatable practice for a public facing library. Best practices of the language should be respected, for the sake of less surprise principle. Hi-jacking a language concept for benefiting of one of its side feature while not actually implementing the concept should be reserved to private code (if not avoided at all).
(Of course since NHibernate comes from Hibernate, many Java practices potentially contradictory with .Net ones have been left. But still, better avoid adding more unconventional practices.)